syncopy/preproc/firws.py

```diff
@@ -0,0 +1,237 @@
+    f_c : float or array_like
+       Cut-off frequenc(ies) in sampling units,
+       maximum is Nyquist `f_c=0.5`. For band-pass
+       and band-stop filters they have to be ordered low to high.
+    filter_type : {'lp', 'hp', 'bp, 'bs'}, optional
+        Select type of filter, either low-pass `'lp'`,
+        high-pass `'hp'`, band-pass `'bp'` or band-stop (Notch) `'bs'`.
+    """
+
+    # order has to be even
+    if order % 2 != 0:
+        order += 1
+
+    if filter_type == 'lp':
+        kernel = windowed_sinc(window, order, f_c)
+        return kernel
+
+    elif filter_type == 'hp':
+        lp_kernel = windowed_sinc(window, order, f_c)
+        kernel = invert_sinc(lp_kernel)
+        return kernel
+
+    if filter_type == 'bp':
+        # high-pass freq is lower than low-pass freq
+        # for band-pass filters
+        f_hp, f_lp = f_c
+    elif filter_type == 'bs':
+        # high-pass freq is higher than low-pass freq
+        # for band-stop filters
+        f_lp, f_hp = f_c
+
+    # construct band filters
+    lp_kernel = windowed_sinc(window, order, f_lp)
+    kernel = windowed_sinc(window, order, f_hp)
+    hp_kernel = invert_sinc(kernel)
+    kernel = lp_kernel + hp_kernel
+
+    # subtract dc component from filter addition
+    # for band-pass filters in the time-domain
+    if filter_type == 'bp':
+        kernel[len(kernel) // 2] -= 1
+
+    return kernel
+
+
+def windowed_sinc(window, order, f_c):
+
+    """
+    Construct the symmetric windowed sinc filter
+    with a cut-off frequency `f_c`
+
+    Parameters
+    ----------
+    window : str
+        One of `scipy.signal.windows`
+    order : int
+       The order of the filter, has to be strictly even
+    f_c : float
+       Cut-off frequency in sampling units,
+       maximum is Nyquist `f_c=0.5`
+
+    Returns
+    -------
+    kernel : :class:`numpy.ndarray`
+        The windowed filter with length `order + 1`
+
+    """
+
+    # angular cut-off frequency
+    omega_c = 2 * np.pi * f_c
+
+    win_func = getattr(sci_win, window)
+    win = win_func(order + 1)
+
+    # one-sided support
+    m_half = np.arange(1, order / 2 + 1)
+    kernel = np.sin(omega_c * m_half) / m_half
+    kernel = np.hstack([kernel[::-1], omega_c, kernel]) * win
+    # normalize to unity gain
+    kernel = kernel / kernel.sum()
+
+    return kernel
+
+
+def invert_sinc(kernel):
+
+    """
+    In frequency space the high-pass filter
+    is just 1 - low-pass. Hence, formally we can
+    calculate the high-pass version
+    in the time domain via the correspinding inverse Fourier:
+
+    Fourier^-1 {1 - Fourier(kernel)} = Fourier^-1(1) - kernel
+
+    This gives a delta-peak at the 0-lag midpoint
+    minus the original low-pass impulse response.
+    In the DSP world the delta distribution corresponds to a mere `1`.
+    """
+    kernel = -kernel
+    # kernel size is always odd
+    kernel[len(kernel) // 2] += 1
+    return kernel
+
+
+def minphaserceps(fkernel):
```

Comment:
What does the `rceps` part of the function name stand for?

Reply:
Could this be a typo, and it should be `minphaserspec` for minimum phase real spectrum?

Reply:
https://en.wikipedia.org/wiki/Cepstrum